src/xpu/sycl/buffer_memory_storage.cpp

```diff
@@ -124,6 +124,7 @@ std::unique_ptr<memory_storage_t> buffer_memory_storage_t::clone() const {
 
     storage->buffer_ = buffer_;
     storage->base_offset_ = base_offset_;
+    storage->offset_ = this->offset_;
```

Comment:
Can we use `offset()` method to get it without moving `offset_` filed from private to protected?

@rjoursler, you seem to used it in the RNN implementation. Can you please clarify the semantics of the `offset_` in memory storage? To me, it seems that the offset is meant to be used for splitting a memory storage into separate logical parts depending on how/where it's being used so carrying over the offset when cloning the memory storage seems weird because the meaning of the offset is context dependent.

Reply:
The meaning used here is that offset is always relative to the base memory object. The reason for this is to align with subbuffer behaviors in the runtimes. For example  [in OpenCL](https://registry.khronos.org/OpenCL/sdk/3.0/docs/man/html/clCreateSubBuffer.html),

> buffer must be a valid buffer object and cannot be a sub-buffer object.

And [in SYCL](https://github.khronos.org/SYCL_Reference/iface/buffer.html)

> b is the buffer with the real data, which must not be a sub-buffer.

Reply:
> The meaning used here is that offset is always relative to the base memory object.

Why can't we use a sub-storage for that?

Reply:
Are you suggesting to use sub-buffers?
From my experience using sub-buffers is not as simple as using an offset. The main issue I have faced is that some backends require sub-buffers to be aligned to some size larger than the buffer's data type. From the OpenCL link above:
```quote
CL_MISALIGNED_SUB_BUFFER_OFFSET if there are no devices in context associated with buffer for which the origin field of the cl_buffer_region structure passed in buffer_create_info is aligned to the CL_DEVICE_MEM_BASE_ADDR_ALIGN value.
```

Reply:
> Why can't we use a sub-storage for that?

Potentially, we could use sub-storage for that, but it would require some changes. Currently, the `get_sub_storage` interface is synonymous with generating sub-buffer runtime objects and, as @Rbiessy mentioned, the problem is that these sub-buffers have restrictions on nesting and alignment. Due to the nature of RNN, it is common to need different ways of splitting the memory into sub-objects. For example, the workspace which corresponds to a single memory allocation by our API actually corresponds to multiple buffers such as gates and states. In addition, each of those buffers may be further subdivided into cells or layers based on what computation is being performed. 

To handle this, we need a more robust interface which allows arbitrary data alignment, ensures no nested sub-buffers occur, and give user control over when sub-buffers are created. The [sub_buffer_t](https://github.com/intel-innersource/libraries.performance.math.onednn/blob/40f939c08dacddf2c863fbabd2806089a4438098/src/gpu/intel/ocl/rnn/rnn_utils.hpp#L384-L419) structure I introduced in `gpu/intel/ocl/rnn/rnn_utils.hpp` (and which is duplicated by this PR), was intended to handle this complexity within the Intel GPU RNN implementation. As this implementation needs similar support, merging the functionality into `memory_storage_t` would make sense.

Reply:
> To handle this, we need a more robust sub_storage interface which allows arbitrary data alignment, ensures no nested sub-buffers occur, and gives control over when sub-buffers are created.

`memory_storage_t::get_sub_storage()` takes an offset and creates a a sub-storage that is essentially a sub-buffer that is always [created from the root buffer](https://github.com/oneapi-src/oneDNN/blob/0fd3b73a25d11106e141ddefd19fcacc74f8bbfe/src/xpu/sycl/buffer_memory_storage.cpp#L108) so we don't create nesting sub-buffers. As for the alignment issue, I'm no sure I understand it, when do we run into the issue?

Reply:
I think I need to clarify the restriction I was referring to apply to OpenCL sub-buffers (and SYCL sub-buffers that may use OpenCL under the hood) which are created when using `memory_storage_t::get_sub_storage()`. On PVC for instance querying `CL_DEVICE_MEM_BASE_ADDR_ALIGN` returns 1024 bits. This means that if for instance the root buffer is at address 0x00000400 (aligned to 1024 already) the only possible offsets are multiples of 1024 bits i.e. multiples of 32 elements when using fp32. This gets more complicated if the root buffer is not aligned and when supporting more data types. In general it is much easier to track the offset as a separate variable.

I agree with the first comment to use a getter method and Anton is planning to look into merging the functionality of `sub_buffer_t` into the main `memory_storage_t` object.

Reply:
>  when do we run into the issue?

All of the calls generating `sub_buffer_t`  objects from the `user_data_t` structure [here](https://github.com/ShanoToni/oneDNN/blob/746808b876c8d3dac0de7e608f80196edcc0b3b5/src/gpu/generic/sycl/rnn/rnn_utils.hpp#L208-L283) may be unaligned depending on the layout and problem dimensions submitted to oneDNN.